### ISSUE TITLE
eq_op false positive

### ISSUE THREAD

**Issue by lukebitts** [Manishearth/rust-clippy#1466]
Consider the following function:

```
fn nan_test() -> bool {
	let v = std::f32::NAN;

	v != v //warning: equal expressions as operands to `!=`
}
```

The `eq_op` warning is wrong here since floats only implement `PartialOrd` and doing `v != v` allows me to detect them in a generic context:

```
fn max<T: PartialOrd>(v1: T, v2: T) -> T {
	if v2 != v2 { v2 }
	else if v2 > v1 { v2 } else { v1 }
}
```

Maybe a solution is to check whether the result of the expression returns something that implements `Ord` or not?

**Comment by llogiq** [comment]
I'd rather specialize reporting for floats and suggest `v.is_nan()` instead.

**Comment by lukebitts** [comment]
I can't use `is_nan` in a generic context. I could bind the parameter to `num::Float` though.

**Comment by llogiq** [comment]
True, but that's a fairly niche use case. Maybe it's best to simply `#[allow(eq_op)]` in your `max` function?

**Comment by lukebitts** [comment]
Actually `num::Float` and `is_nan` turned out to be a better abstraction for my project. So I guess the warning was right

**Comment by BusyJay** [comment]
#6167 makes `eq_op` lint on assert families, but when users choose to implement PartialEq for a custom type, they may use `assert_eq` to test if it's correctly implemented, which leads to the lint error.
